File: qa-automation/src/pages/AuthPage.js
const BasePage = require('./BasePage');
const logger = require('../utils/Logger');

class AuthPage extends BasePage {
    // ----------------------------------------------------------------------
    // Selectors
    // ----------------------------------------------------------------------
    get container() { return $('.auth-container'); }
    get heading() { return $('.auth-header h2'); }
    get roleLabel() { return $('.auth-header .text-muted'); }
    get backBtn() { return $('.back-btn'); }
    
    // Form Inputs
    get inputEmail() { return $('input[type="email"]'); }
    get inputPassword() { return $('input[type="password"]'); }
    get inputFullName() { return $('input[type="text"][placeholder*="Doe"]'); }
    get inputMobile() { return $('input[type="tel"]'); }
    get inputAddress() { return $('input[placeholder*="Clinical"]'); }
    get selectGender() { return $('select'); }
    get inputAvatar() { return $('input#avatarUpload'); }

    // Actions & Messaging
    get btnSubmit() { return $('button[type="submit"]'); }
    get linkToggleMode() { return $('.text-link'); }
    get errorMessage() { return $('.error-message'); }

    // ----------------------------------------------------------------------
    // Methods
    // ----------------------------------------------------------------------

    /**
     * Opens the Auth page for a given role
     * @param {string} role 'doctor' or 'patient'
     */
    async open(role = 'doctor') {
        logger.info(`Opening Auth Page for role: ${role}`);
        await super.open(`auth?role=${role}`);
        await this.waitForElement(this.container);
    }

    /**
     * Toggles between Login and Signup modes
     */
    async toggleAuthMode() {
        logger.info('Toggling Authentication mode (Login <-> Signup)');
        await this.clickElement(this.linkToggleMode);
    }

    /**
     * Performs a standard login
     * @param {string} email
     * @param {string} password
     */
    async login(email, password) {
        logger.info(`Performing login for email: ${email}`);
        await this.typeText(this.inputEmail, email);
        await this.typeText(this.inputPassword, password);
        await this.clickElement(this.btnSubmit);
    }

    /**
     * Fills out the signup form completely
     */
    async fillSignupForm({ fullName, mobile, address, gender, email, password }) {
        logger.info(`Filling signup form for: ${email}`);
        if (fullName) await this.typeText(this.inputFullName, fullName);
        if (mobile) await this.typeText(this.inputMobile, mobile);
        if (address) await this.typeText(this.inputAddress, address);
        
        if (gender) {
            await this.waitForElement(this.selectGender);
            await this.selectGender.selectByAttribute('value', gender);
        }
        
        if (email) await this.typeText(this.inputEmail, email);
        if (password) await this.typeText(this.inputPassword, password);
    }

    /**
     * Submits the signup form
     */
    async submitSignup() {
        logger.info('Submitting signup form');
        await this.clickElement(this.btnSubmit);
    }

    /**
     * Gets the current error message text
     */
    async getErrorMessage() {
        await this.waitForElement(this.errorMessage, 5000);
        return this.errorMessage.getText();
    }
    
    /**
     * Checks if error message is displayed
     */
    async isErrorDisplayed() {
        return this.errorMessage.isDisplayed();
    }
}

module.exports = new AuthPage();

File: qa-automation/appium/pages/auth_page.py
from .base_page import BaseMobilePage
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import Select
import logging

logger = logging.getLogger("AppiumPython")

class MobileAuthPage(BaseMobilePage):
    # Selectors
    CONTAINER = (By.CSS_SELECTOR, '.auth-container')
    HEADING = (By.CSS_SELECTOR, '.auth-header h2')
    ROLE_LABEL = (By.CSS_SELECTOR, '.auth-header .text-muted')
    BACK_BTN = (By.CSS_SELECTOR, '.back-btn')
    
    INPUT_EMAIL = (By.CSS_SELECTOR, 'input[type="email"]')
    INPUT_PASSWORD = (By.CSS_SELECTOR, 'input[type="password"]')
    INPUT_FULL_NAME = (By.CSS_SELECTOR, 'input[placeholder*="Doe"]')
    INPUT_MOBILE = (By.CSS_SELECTOR, 'input[type="tel"]')
    INPUT_ADDRESS = (By.CSS_SELECTOR, 'input[placeholder*="Clinical"]')
    SELECT_GENDER = (By.CSS_SELECTOR, 'select')
    INPUT_AVATAR = (By.CSS_SELECTOR, 'input#avatarUpload')
    
    BTN_SUBMIT = (By.CSS_SELECTOR, 'button[type="submit"]')
    LINK_TOGGLE_MODE = (By.CSS_SELECTOR, '.text-link')
    ERROR_MESSAGE = (By.CSS_SELECTOR, '.error-message')

    def login(self, email, password):
        """Authenticates the user in the mobile WebView"""
        logger.info(f"Appium: Typing email: {email}")
        self.type_text(*self.INPUT_EMAIL, email)
        
        logger.info("Appium: Typing password")
        self.type_text(*self.INPUT_PASSWORD, password)
        
        logger.info("Appium: Tapping Sign In")
        self.click_element(*self.BTN_SUBMIT)

    def toggle_auth_mode(self):
        """Toggles between signup and login screens"""
        logger.info("Appium: Tapping Toggle Mode link")
        self.click_element(*self.LINK_TOGGLE_MODE)

    def get_error_message(self):
        """Retrieves validation error message from screen"""
        element = self.wait_for_element(*self.ERROR_MESSAGE)
        return element.text

    def fill_signup_form(self, full_name=None, mobile=None, address=None, gender=None, email=None, password=None):
        """Fills out registration details"""
        logger.info("Appium: Filling mobile signup form details")
        if full_name:
            self.type_text(*self.INPUT_FULL_NAME, full_name)
        if mobile:
            self.type_text(*self.INPUT_MOBILE, mobile)
        if address:
            self.type_text(*self.INPUT_ADDRESS, address)
        if gender:
            select_el = self.wait_for_element(*self.SELECT_GENDER)
            Select(select_el).select_by_value(gender)
        if email:
            self.type_text(*self.INPUT_EMAIL, email)
        if password:
            self.type_text(*self.INPUT_PASSWORD, password)
